### DISCUSSION TITLE
Firefox addon has been removed from Mozilla site

### DISCUSSION THREAD

**Discussion by NealWilxite** [ActivityWatch/activitywatch#818] [Q&A] [answered]
**Addon page**

![image](https://user-images.githubusercontent.com/54668805/198983068-715f26e1-c398-442f-ad33-9bd28d1ef035.png)

**Search results**

![image](https://user-images.githubusercontent.com/54668805/198984194-11515af2-64d5-46d8-953a-e8be2df9a71e.png)

Is there an up to date version on github. I could only find an archived one that's 7 years old (https://github.com/ActivityWatch/aw-watcher-firefox).

**Comment by ErikBjare** [comment, marked as answer]
Thanks for reporting, I was not aware! 

Looks like Mozilla decided to pull it out of the blue? Logged in and saw this:

<img width=500 alt="image" src="https://user-images.githubusercontent.com/1405370/198991396-42201826-3ac3-4270-b901-b6f4f58ed184.jpg">

As you can see, I replied and hope they relist it soon. 

The correct repo is: https://github.com/ActivityWatch/aw-watcher-web

Also reported here: https://github.com/ActivityWatch/aw-watcher-web/issues/94

Edit: It has now been 16 days without reply from Mozilla...

Edit 2: Made a post on their forums: https://discourse.mozilla.org/t/add-on-disabled-after-1y-16-days-without-reply-action/107797

Edit 3: After much back and forth with Mozilla's incompetent reviewers who can't follow simple instructions, they wrote the following, with my reply:

<img width="658" alt="image" src="https://user-images.githubusercontent.com/1405370/204296244-0a298519-51c7-4739-845f-1f460728771e.png">

The links:
 - https://extensionworkshop.com/documentation/publish/add-on-policies/#data-disclosure-collection-and-management
 - https://extensionworkshop.com/documentation/develop/best-practices-for-collecting-user-data-consents/

I have no interest in satisfying these pointless rules, which provide no user protection at all.

If anyone wants to do the work, be my guest, but I have no more patience with the Mozilla Addons people.

**Reply by ErikBjare** [reply]
I've now updated the answer with my (final?) response.

Those who want to deal with Mozilla Addons are welcome to update the extension, create your own listing, and get it approved. 

Only then will I touch this/talk to Mozilla Addons again.

**Reply by ErikBjare** [reply]
~~I've uploaded the last Mozilla-signed xpi here: https://github.com/ActivityWatch/aw-watcher-web/releases/tag/v0.4.3~~

~~It should be possible to install without having to manually build and download Firefox Nightly / Developer Edition.~~

Edit: Apparently I wasn't allowed to do that, see comment below.

**Reply by qimmiq** [reply]
thanks, for me it works ;)

**Reply by ErikBjare** [reply]
@qimmiq That's great to hear! 

Will link to it directly from the README too. 

Edit: I've now updated the [README](https://github.com/ActivityWatch/aw-watcher-web/blob/master/README.md) with workarounds. Thanks for reporting everyone!

**Reply by ErikBjare** [reply]
Mozilla Addons' reviewers, in their infinite stupidity, today sent me the following:

>Self-distributing a version that we rejected due to policy violations is circumventing our review process and our policies. We ask that you stop self-hosting rejected and/or violating versions immediately (by Friday, Dec 2). Otherwise we might have to block your add-on from running in Firefox.

As such, **I've had to remove the Mozilla-signed xpi from the release linked above.**

My reply to them:

>As much as I'd love to explain to my users how your ridiculous your request/policy is when their shit breaks for truly the dumbest of reasons, I've removed the "self-distributed" xpi.
>
>As previously stated, I will not be submitting a new version until you fix your policy, or the way in which you apply it. (expected time: never)
>
>To clarify one last time what my issue is: Asking for "consent" in a strictly local-only context, where the user has *very explicitly and intentionally* taken steps to do exactly what a consent dialog would ask for is beyond "do you accept cookies?"-banner stupid. I refuse to write software that succumbs to such stupidity. 
>
>Repeated incompetence during an unnecessarily drawn out review process is just the icing on the cake.
>
>You've lost a Mozilla fan for life. I didn't want to believe in the supposed "slow demise of Mozilla" I keep hearing about, but now I do :(

Yes, I'm both pissed and sad about this.

You options are:
 - Switch browser (easiest, and frankly recommended at this point)
 - Use Firefox Developer Edition & build/install an unsigned version of the extension (see README)
 - Ask an existing user to send you the Mozilla-signed xpi (could ruin things for existing users if Mozilla decides to fulfill their threat to blacklist the extension due to "self-distribution", so please don't)

**Reply by renyuneyun** [reply]
The policy seems quite rough... That should not be an enforcement for all addons, in particular if the accompanied application is a local/standalone application AND FLOSS. It seems either the reviewer or the policymaker has no idea what a standalone application is... Quite a shame for an employee of an organization like Mozilla.
And requiring no distribution of addon from custom route is also ridiculous...

But I can see how it can make sense for AW, in particular when the user wants to (temporarily) disable tracking.
I'm not an expert in this topic either (for now). But maybe simply adding a "pop-up window" or one setting item, and have an "enable/disable" toggle would suffice?

Re privacy policy: I faced a similar situation for something else (also a standalone but information-intensive application, for Android), though in my case it's not mandatory. I eventually wrote a (bad) poem as the privacy "statement", rather than "policy". Not sure if it works for this case or not.

**Reply by johan-bjareholt** [reply]
> But maybe simply adding a "pop-up window" or one setting item, and have an "enable/disable" toggle would suffice?

There already is an button to temporarily disable the watcher which is very easy to access.

So I guess the only missing part they require is that upon first install, they require some setup wizard or something that you have to check a checkbox saying "Yes I agree that this watcher transmits data to my own computer" or something. So incredibly silly, but doable.

**Reply by flexagoon** [reply]
@johan-bjareholt While this is, indeed, a weird requirement in this case, various platforms have it pretty often. Google Play, AFAIK, also requires all apps to have a privacy policy, even if they don't transmit any data. You can see an example of privacy policy for a fully offline app here

https://florisboard.org/legal/privacy/

So you just need to write a short privacy policy saying "we collect your visited sites, but that information is stored locally".

Then you just show a tab on first install which has that information and then two buttons: "I agree" and "uninstall the extension". An example of such dialog can be found here

https://extensionworkshop.com/documentation/develop/best-practices-for-collecting-user-data-consents

(See the "Required data collection" example at the very bottom of a page)

**Reply by flexagoon** [reply]
Here's another example of a privacy policy for something that doesn't transmit user data - uBlock Origin.

https://github.com/gorhill/uBlock/wiki/Privacy-policy

As you can see, it doesn't even have to be formal, it just has to clearly explain what data the extension collects and that it doesn't send it anywhere besides your local activitywatch server.

**Reply by johan-bjareholt** [reply]
@flexagoon Those two examples are not directly comparable, because

- Florisboard actually transmits data to a remote server (even if it is not stored) and therefore has a legit and useful privacy policy.
- uBlock origin transmits no data at all, and therefore Mozilla does not require an acknowledgement from the user upon install.

Our case is that we transmit data, but only to the local host. So you could say that our case is in between Florisboard and uBlock.

Also the issue is not that we need to have a privacy policy, it's that Mozilla require us to have some kind of user acknowledgement upon install. And as I said before is possible to fix, it's just that no one has taken the time to implement it.

**Reply by flexagoon** [reply]
@johan-bjareholt 

> Florisboard actually transmits data to a remote server (even if it is not stored) and therefore has a legit and useful privacy policy.

Florisboard does not transmit any data to a remote server.

> Also the issue is not that we need to have a privacy policy, it's that Mozilla require us to have some kind of user acknowledgement upon install. And as I said before is possible to fix, it's just that no one has taken the time to implement it.

So you don't actually need to have a privacy policy, you just need to add a user acknowledgement dialog? If that's the case, I can probably try and implement this.

**Reply by vaportail** [reply]
Did these [censored] forbid you from even distributing the built extension?
Yeah, we desperately need another browser. Mozilla is getting as bad as Google, especially with fishy policy decisions and a new-found love for web censorship.
I've got no words to describe how idiotic these decisions are.

Sounds more like some proprietary paid solution **paid off @Mozilla reviewers** to block competition. If so that has to be investigated.
But that's the only explanation I can come up with other than someone in Mozilla suffering from a massive stroke.

**Reply by johan-bjareholt** [reply]
> Florisboard does not transmit any data to a remote server.

Oh sorry for misunderstanding.

> So you don't actually need to have a privacy policy, you just need to add a user acknowledgement dialog? If that's the case, I can probably try and implement this.

That would be highly appreciated if you did!

**Reply by johan-bjareholt** [reply]
@vaportail They did not forbid us from distributing a built extension, they forbid us to distribute extensions **signed with mozillas keys**.

IMO their decision makes sense from Mozillas point of view, they are improving the transparency of what data extensions transmit. Since ActivityWatch is quite an odd extension in the sense that we transmit data but only to localhost, they probably didn't have that in mind when they made this policy so it's possibly more strict than necessary for our extension in particular.

**Reply by Sleepful** [reply]
Sad... sooo who wants to code the "consent dialog"? 🤪

**Reply by flexagoon** [reply]
@Sleepful as I said, I can try to do this when I have the time. Although I don't have a lot of experience with browser extensions, so I can't guarantee anything, but this doesn't seem too hard.

**Reply by Sleepful** [reply]
Nice! I got it working for now, just had to replace Firefox with `Firefox Developer Edition` and then build from source, rather simple.

**Reply by vaportail** [reply]
@johan-bjareholt oh I agree, from their point of view it makes sense. It is their point of view I have a problem with.
As for the policy itself, it will be as useful as GDPR cookie consent popups. At least someone will be able to pat themselves on the back thinking they did something good.

I can understand them not wanting something they signed distributed - but the bottom line is people can't use tools they want to use. That's a closed ecosystem controlled by people, recently people with a fishy agenda. Which is why I say they are no better than Google. That and hidden telemetry.

**Reply by Morpheus0x** [reply]
I am currently working on a pull request for adding support for [issue #82](https://github.com/ActivityWatch/aw-watcher-web/issues/82) since this is something I need. I have multiple Firefox profiles open at the same time, and all of them should be tracked with activity watch. 
But since I just noticed this issue now, I will start working on that, since this is basically required for me (never going to use anything but Firefox with all telemetry disabled).

@ErikBjare I can totally understand your frustration, but think it is very important to get that original extension back up so people who have the extension still installed will get updates in the future.

I am going to implement the consent form with an option to use [enterprise policy](https://extensionworkshop.com/documentation/enterprise/adding-policy-support-to-your-extension/) to pre-accept it. Does anyone know how this would behave in chrome? Is the manifest permission `storage` and the `browser.storage.managed` API the same as the Firefox docs describe it?

If someone else could write (or copy) a "no-data-collection" privacy policy, we can have the extension back up soon.

Edit: Also, I would like the current enable setting and the baseURL to be able to be set via enterprise policy as well. Should this be a separate pull request, or is it OK if I submit this all in one (Since I am adding policy support anyway)?

Edit 2: just realised that baseURL should basically never be changed, I was thinking about using a central server to do my logging but since network access isn't always guaranteed, I think I will keep this as is and only use the policy to consent to "data collection"

**Reply by Morpheus0x** [reply]
Pull Request is [here](https://github.com/ActivityWatch/aw-watcher-web/pull/99)

**Reply by ErikBjare** [reply]
I've submitted a new version with the privacy notice & consent dialog. Waiting for review by Mozilla Addons now 🤞

**Reply by omniraptorr** [reply]
any updates on this? From what I can tell it doesn't usually take them longer than a couple days to review.

**Reply by johan-bjareholt** [reply]
@omniraptorr It was unfortunately rejected again. Have not had time to dig into the details yet.

![Screenshot_2023-03-19-22-55-14-42](https://user-images.githubusercontent.com/1398500/226212069-b052222b-2933-4ec0-91cc-5533d8e29367.jpg)

**Reply by takecare** [reply]
i think they mean it's not acceptable to have an enterprise policy that makes it possible for the extension to "auto-accept" its own consent. so the check

```js
try {
  if (await browser.storage.managed.get("consentOfflineDataCollection")) {
    return false
  }
} catch (e) {
  console.error('managed storage error: ', e)
  return true
}
```

should be removed (alongside the policy instructions in the readme). let me know if you want me to pr it.

**Reply by ErikBjare** [reply]
I've just resubmitted a version with the managed storage pre-consent feature disabled.

**Reply by johan-bjareholt** [reply]
Finally, the extension got approved!!!

**Reply by tobiasdiez** [reply]
Yeah, thanks a lot for your continued work on this!

For reference, here it is: https://addons.mozilla.org/en-US/firefox/addon/aw-watcher-web/

**Comment by roke-julian-lockhart** [comment]
Can we convert this to an issue? I'd like to be able to be notified solely of its remediation, rather than be subscribed for every update.

**Reply by flexagoon** [reply]
What do you want to be notified about? This is already solved, the addon is available on AMO

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/ActivityWatch/discussions/818#discussioncomment-7084091

Sorry, @flexagoon, couldn't find it for whatever reason. Found it now. Apologies.
